File: ios/HybridPedometer.swift
import Foundation
import CoreMotion
import NitroModules

class HybridPedometer: HybridPedometerSpec {

  // MARK: - Private

  private let pedometer = CMPedometer()
  private var _isActive = false

  // MARK: - Properties

  var isAvailable: Bool {
    return CMPedometer.isStepCountingAvailable()
  }

  var isActive: Bool {
    return _isActive
  }

  var onUpdate: ((PedometerData) -> Void)?

  // MARK: - Permission Methods

  func checkPermission() throws -> Promise<PermissionStatus> {
    return Promise.resolved(withResult: self.mapAuthorizationStatus(CMPedometer.authorizationStatus()))
  }

  func requestPermission() throws -> Promise<PermissionStatus> {
    return Promise.async { [weak self] () -> PermissionStatus in
      guard let self = self else { return .denied }

      // CMPedometer doesn't have a standalone "request" API.
      // Permission is triggered on first data query.
      // We do a tiny historical query to trigger the system prompt.
      return try await withCheckedThrowingContinuation { continuation in
        let now = Date()
        let oneMinuteAgo = now.addingTimeInterval(-60)

        self.pedometer.queryPedometerData(from: oneMinuteAgo, to: now) { _, error in
          if let error = error as? NSError, error.code == Int(CMErrorMotionActivityNotAuthorized.rawValue) {
            continuation.resume(returning: .denied)
          } else {
            continuation.resume(returning: self.mapAuthorizationStatus(CMPedometer.authorizationStatus()))
          }
        }
      }
    }
  }

  // MARK: - Sensor Methods

  func start() throws {
    guard CMPedometer.isStepCountingAvailable() else {
      throw NSError(
        domain: "NitroSensors",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: "Pedometer is not available on this device"]
      )
    }

    _isActive = true

    pedometer.startUpdates(from: Date()) { [weak self] data, error in
      guard let data = data, error == nil else { return }

      let sensorData = PedometerData(
        steps: Double(truncating: data.numberOfSteps),
        distance: data.distance?.doubleValue ?? -1,
        currentPace: data.currentPace?.doubleValue ?? -1,
        currentCadence: data.currentCadence?.doubleValue ?? -1,
        floorsAscended: data.floorsAscended?.doubleValue ?? -1,
        floorsDescended: data.floorsDescended?.doubleValue ?? -1,
        timestamp: Date().timeIntervalSinceReferenceDate
      )

      self?.onUpdate?(sensorData)
    }
  }

  func stop() throws {
    pedometer.stopUpdates()
    _isActive = false
  }

  func queryHistoricalData(startTime: Double, endTime: Double) throws -> Promise<PedometerData> {
    return Promise.async { [weak self] () -> PedometerData in
      guard let self = self else {
        throw NSError(domain: "NitroSensors", code: 2, userInfo: [NSLocalizedDescriptionKey: "Pedometer instance deallocated"])
      }

      // Convert Unix milliseconds to Date
      let start = Date(timeIntervalSince1970: startTime / 1000.0)
      let end = Date(timeIntervalSince1970: endTime / 1000.0)

      return try await withCheckedThrowingContinuation { continuation in
        self.pedometer.queryPedometerData(from: start, to: end) { data, error in
          if let error = error {
            continuation.resume(throwing: error)
            return
          }

          guard let data = data else {
            continuation.resume(throwing: NSError(
              domain: "NitroSensors",
              code: 3,
              userInfo: [NSLocalizedDescriptionKey: "No pedometer data returned"]
            ))
            return
          }

          let result = PedometerData(
            steps: Double(truncating: data.numberOfSteps),
            distance: data.distance?.doubleValue ?? -1,
            currentPace: data.currentPace?.doubleValue ?? -1,
            currentCadence: data.currentCadence?.doubleValue ?? -1,
            floorsAscended: data.floorsAscended?.doubleValue ?? -1,
            floorsDescended: data.floorsDescended?.doubleValue ?? -1,
            timestamp: Date().timeIntervalSinceReferenceDate
          )

          continuation.resume(returning: result)
        }
      }
    }
  }

  // MARK: - Helpers

  private func mapAuthorizationStatus(_ status: CMAuthorizationStatus) -> PermissionStatus {
    switch status {
    case .authorized:
      return .granted
    case .denied:
      return .denied
    case .restricted:
      return .restricted
    case .notDetermined:
      return .denied
    @unknown default:
      return .denied
    }
  }
}

File: android/src/main/java/com/margelo/nitro/nitrosensorskit/HybridPedometer.kt
package com.margelo.nitro.nitrosensorskit

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Build
import androidx.annotation.Keep
import androidx.core.content.ContextCompat
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise

@Keep
@DoNotStrip
class HybridPedometer : HybridPedometerSpec(), SensorEventListener {

  // MARK: - Private

  private val context: Context
    get() = NitroModules.applicationContext ?: throw Error("No ApplicationContext set!")

  private val sensorManager: SensorManager
    get() = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager

  private val stepCounterSensor: Sensor?
    get() = sensorManager.getDefaultSensor(Sensor.TYPE_STEP_COUNTER)

  private var _isActive = false

  // TYPE_STEP_COUNTER gives cumulative steps since boot.
  // We track the initial value to compute relative steps since start().
  private var initialStepCount: Float? = null

  // MARK: - Properties

  override val isAvailable: Boolean
    get() = stepCounterSensor != null

  override val isActive: Boolean
    get() = _isActive

  override var onUpdate: ((data: PedometerData) -> Unit)? = null

  // MARK: - Permission Methods

  override fun checkPermission(): Promise<PermissionStatus> {
    return Promise.resolved(getCurrentPermissionStatus())
  }

  override fun requestPermission(): Promise<PermissionStatus> {
    // Android runtime permissions must be requested from an Activity.
    // From a library context without Activity access, we can only check.
    // The consuming app is responsible for requesting ACTIVITY_RECOGNITION
    // before calling start(). We return the current status.
    //
    // Pre-API 29: no runtime permission needed, always granted.
    return Promise.resolved(getCurrentPermissionStatus())
  }

  // MARK: - Sensor Methods

  override fun start() {
    val s = stepCounterSensor
      ?: throw IllegalStateException("Pedometer is not available on this device")

    initialStepCount = null // Reset on each start
    sensorManager.registerListener(this, s, SensorManager.SENSOR_DELAY_NORMAL)
    _isActive = true
  }

  override fun stop() {
    sensorManager.unregisterListener(this)
    _isActive = false
    initialStepCount = null
  }

  override fun queryHistoricalData(startTime: Double, endTime: Double): Promise<PedometerData> {
    // Android TYPE_STEP_COUNTER doesn't support historical queries.
    // Return an empty result with 0 steps.
    return Promise.resolved(
      PedometerData(
        steps = 0.0,
        distance = -1.0,
        currentPace = -1.0,
        currentCadence = -1.0,
        floorsAscended = -1.0,
        floorsDescended = -1.0,
        timestamp = System.currentTimeMillis().toDouble() / 1000.0
      )
    )
  }

  // MARK: - SensorEventListener

  override fun onSensorChanged(event: SensorEvent?) {
    val e = event ?: return
    val cumulativeSteps = e.values[0]

    // Capture initial step count on first reading
    if (initialStepCount == null) {
      initialStepCount = cumulativeSteps
    }

    val stepsSinceStart = cumulativeSteps - (initialStepCount ?: cumulativeSteps)

    // Android TYPE_STEP_COUNTER only provides step count.
    // Distance, pace, cadence, and floors are not available.
    val data = PedometerData(
      steps = stepsSinceStart.toDouble(),
      distance = -1.0,      // Not available on Android
      currentPace = -1.0,   // Not available on Android
      currentCadence = -1.0, // Not available on Android
      floorsAscended = -1.0, // Not available on Android
      floorsDescended = -1.0, // Not available on Android
      timestamp = e.timestamp.toDouble() / 1_000_000_000.0
    )

    onUpdate?.invoke(data)
  }

  override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {
    // No-op
  }

  // MARK: - Helpers

  private fun getCurrentPermissionStatus(): PermissionStatus {
    // Pre-API 29: ACTIVITY_RECOGNITION permission doesn't exist, always granted
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
      return PermissionStatus.GRANTED
    }

    val result = ContextCompat.checkSelfPermission(
      context,
      Manifest.permission.ACTIVITY_RECOGNITION
    )

    return if (result == PackageManager.PERMISSION_GRANTED) {
      PermissionStatus.GRANTED
    } else {
      // Android doesn't distinguish between "denied" and "not determined"
      // at the library level without Activity access.
      PermissionStatus.DENIED
    }
  }
}
